test/extensions/filters/listener/original_src/original_src_test.cc

```diff
@@ -47,47 +71,93 @@ TEST_F(OriginalSrcTest, onNewConnectionUnixSocketSkips) {
 
   filter->onAccept(callbacks_);
   std::vector<uint8_t> key;
-  option->hashKey(key);
+  // not ideal -- we're assuming that the original_src option is first, but it's a fair assumption
+  // for now.
+  options->at(0)->hashKey(key);
   std::vector<uint8_t> expected_key = {1, 2, 3, 4};
 
   EXPECT_EQ(key, expected_key);
 }
 
 TEST_F(OriginalSrcTest, onNewConnectionIpv4AddressBleachesPort) {
   auto filter = makeDefaultFilter();
-  Network::Socket::OptionConstSharedPtr option;
+  Network::Socket::OptionsSharedPtr options;
   setAddressToReturn("tcp://1.2.3.4:80");
-  Network::Socket::OptionsSharedPtr options{std::make_shared<Network::Socket::Options>()};
-  EXPECT_CALL(callbacks_.socket_, addOption_(_)).WillOnce(SaveArg<0>(&option));
+  EXPECT_CALL(callbacks_.socket_, addOptions_(_)).WillOnce(SaveArg<0>(&options));
 
   filter->onAccept(callbacks_);
 
   NiceMock<Network::MockConnectionSocket> socket;
   const auto expected_address = Network::Utility::parseInternetAddress("1.2.3.4");
   EXPECT_CALL(socket, setLocalAddress(PointeesEq(expected_address)));
 
-  option->setOption(socket, envoy::api::v2::core::SocketOption::STATE_PREBIND);
+  // not ideal -- we're assuming that the original_src option is first, but it's a fair assumption
+  // for now.
+  options->at(0)->setOption(socket, envoy::api::v2::core::SocketOption::STATE_PREBIND);
+}
+
+TEST_F(OriginalSrcTest, filterAddsTransparentOption) {
+  if (!ENVOY_SOCKET_IP_TRANSPARENT.has_value()) {
+    // The option isn't supported on this platform. Just skip the test.
+    return;
+  }
+
+  auto filter = makeDefaultFilter();
+  Network::Socket::OptionsSharedPtr options;
+  setAddressToReturn("tcp://1.2.3.4:80");
+  EXPECT_CALL(callbacks_.socket_, addOptions_(_)).WillOnce(SaveArg<0>(&options));
+
+  filter->onAccept(callbacks_);
+
+  auto transparent_option = findOptionDetails(*options, ENVOY_SOCKET_IP_TRANSPARENT,
+                                              envoy::api::v2::core::SocketOption::STATE_PREBIND);
+
+  EXPECT_TRUE(transparent_option.has_value());
 }
 
+TEST_F(OriginalSrcTest, filterAddsMarkOption) {
+  if (!ENVOY_SOCKET_SO_MARK.has_value()) {
+    // The option isn't supported on this platform. Just skip the test.
+    return;
+  }
+
+  auto filter = makeMarkingFilter(1234);
+  Network::Socket::OptionsSharedPtr options;
+  setAddressToReturn("tcp://1.2.3.4:80");
+  EXPECT_CALL(callbacks_.socket_, addOptions_(_)).WillOnce(SaveArg<0>(&options));
+
+  filter->onAccept(callbacks_);
+
+  auto mark_option = findOptionDetails(*options, ENVOY_SOCKET_SO_MARK,
+                                       envoy::api::v2::core::SocketOption::STATE_PREBIND);
+
+  ASSERT_TRUE(mark_option.has_value());
+  uint32_t value = 1234;
+  absl::string_view value_as_bstr(reinterpret_cast<const char*>(&value), sizeof(value));
+  std::string mark_value_as_string = std::string(value_as_bstr);
+  EXPECT_EQ(mark_value_as_string, mark_option->value_);
```

Comment:
You should be able to compare string and string_view without explicitly construct string above.